src/main/ts/ui/Dialog.ts

```diff
@@ -0,0 +1,137 @@
+        { text: 'Current tree', value: 'currentTree' },
+      ]
+    });
+  }
+
+  return items;
+}
+
+const nestedListDialog = (editor: Editor): PublicDialog.DialogSpec<EnhancedListDialogData> => {
+  const selectedElement = SugarElement.fromDom(editor.selection.getNode());
+
+  return {
+    title: 'Enhanced List',
+    body: {
+      type: 'panel',
+      items: getPanelItems(selectedElement),
+    },
+    buttons: [
+      {
+        type: 'cancel',
+        name: 'cancel',
+        text: 'Cancel'
+      },
+      {
+        type: 'submit',
+        name: 'save',
+        text: 'Save',
+        primary: true
+      }
+    ],
+    onSubmit: (api) => {
+      const data = api.getData();
+
+      if (Utils.isListItemNode(selectedElement) || Utils.isListNode(selectedElement)) {
+        const elements = getElementsToApply(selectedElement, data.itemsToApplyStyle);
+        editor.undoManager.transact(() => {
+          Arr.each((elements), (x) => {
+            if ((/^(LI)$/).test(x.nodeName)) {
+              editor.dom.setStyle(x, 'padding-inline-start', `${data.paddingLeft}px`);
+            } else {
+              editor.dom.setStyle(x, 'list-style-type', `${data.listStyle}`);
+            }
+          });
+        });
+      } else {
+        const listStyle = data.listStyle;
+        if (Arr.find(['circle', 'disc', 'square'], (x) => x === listStyle).isSome()) {
+          editor.execCommand('InsertUnorderedList', false, { 'list-style-type': data.listStyle })
+        } else {
+          editor.execCommand('InsertOrderedList', false, { 'list-style-type': data.listStyle })
+        }
+        editor.dom.setStyle(editor.selection.getNode(), 'padding-inline-start', data.paddingLeft + 'px');
+      }
+
+      api.close();
+    }
+  }
+}
+
+const getElementsToApply = (selElement: SugarElement, appliedTo: string): Node[] => {
+  const parent = Traverse.parent(selElement).getOr(selElement);
+  let elements = [parent];
+
+  if (appliedTo === 'selectedListAndParent') {
+    const parents = PredicateFilter.ancestors(parent, Utils.isListNode);
+    elements = elements.concat(parents);
+  } else if (appliedTo === 'selectedListAndChildren') {
+    const children = PredicateFilter.descendants(parent, Utils.isListNode);
+    elements = elements.concat(children);
+  } else if (appliedTo === 'currentTree') {
+    const children = PredicateFilter.descendants(parent, Utils.isListNode);
+    const parents = PredicateFilter.ancestors(parent, Utils.isListNode);
+
+    elements = elements.concat(parents);
+    elements = elements.concat(children);
+  }
+
+  const childElements = getChildElements(elements);
+  elements = elements.concat(childElements)
+  return Arr.map(elements, (x) => x.dom);
+}
+
+const getChildElements = (elements: SugarElement<Node>[]): SugarElement<Node>[] => {
+  return Arr.flatten(
+    Arr.map(elements, (x) => PredicateFilter.children(x, Utils.isListItemNode))
+  );
```

Comment:
```suggestion
  return Arr.bind(elements, (x) => PredicateFilter.children(x, Utils.isListItemNode));
```
`Arr.bind` is an alias for flatten => map.

Reply:
Yeah. If you find yourself wanting to build together a heap of arrays, and those arrays were the result of a function, then use `Arr.bind`

Interestingly, I would have thought that the traversal order would be different between `Arr.bind` and `flatten(map(..))` (as in `Arr.bind` would process concatenating the result of each function to the accumulated total before running f on the next item), but Mitch is correct ... it is just an alias. Interesting.